_posts/2019-09-30-bert-r/bert_r.Rmd

```diff
@@ -0,0 +1,536 @@
+
+```{r}
+k_bert = import('keras_bert')
+token_dict = k_bert$load_vocabulary(vocab_path)
+tokenizer = k_bert$Tokenizer(token_dict)
+```
+
+### How tokenizer works?
+
+BERT uses a WordPiece tokenization strategy. If a word is Out-of-vocabulary (OOV), then BERT will break it down into subwords. (eating => eat, ##ing).
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = "[BERT input representation. The input embeddings are the sum of the token embeddings, the segmentation embeddings, and the position embeddings](https://arxiv.org/pdf/1810.04805.pdf)"}
+knitr::include_graphics("images/emb.png")
+```
+
+## Define model parameters and column names
+
+Similar to Keras standard parameters: input length, batch size, epochs and learning rate should be defined for training a BERT model with Keras.
+
+```{r}
+seq_length = 50L
+bch_size = 70
+epochs = 1
+learning_rate = 1e-4
+
+DATA_COLUMN = 'comment_text'
+LABEL_COLUMN = 'target'
+```
+
+> Note: the max input length is 512 and it is extremely computational even for GPU. 
+
+## Load BERT model into R
+
+We can load the BERT model and automatically pad sequences with ```seq_len``` function. This is why using keras-bert^[[Implementation of the BERT. Official pre-trained models could be loaded for feature extraction and prediction](https://github.com/CyberZHG/keras-bert)] is a better choice, it makes the loading process very easy and comfortable.
+
+```{r}
+model = k_bert$load_trained_model_from_checkpoint(
+  config_path,
+  checkpoint_path,
+  training=T,
+  trainable=T,
+  seq_len=seq_length)
+
+```
+
+## Data structure, reading, preparation
+
+The dataset for this post is taken from [Kaggle Jigsaw Unintended Bias in Toxicity Classification competition](https://www.kaggle.com/c/jigsaw-unintended-bias-in-toxicity-classification).
+
+In order to prepare the dataset, we should write a preprocessing function which will read and tokenize data simultaneously. Then, we feed the outputs of the function as input for BERT model.
+
+```{r}
+# tokenize text
+tokenize_fun = function(dataset) {
+  c(indices, target, segments) %<-% list(list(),list(),list())
+  for ( i in 1:nrow(dataset)) {
+    c(indices_tok, segments_tok) %<-% tokenizer$encode(dataset[[DATA_COLUMN]][i], 
+                                                       max_len=seq_length)
+    indices = indices %>% append(list(as.matrix(indices_tok)))
+    target = target %>% append(dataset[[LABEL_COLUMN]][i])
+    segments = segments %>% append(list(as.matrix(segments_tok)))
+  }
+  return(list(indices,segments, target))
+}
+```
+
+```{r}
+# read data
+dt_data = function(dir, rows_to_read){
+  data = data.table::fread(dir, nrows=rows_to_read)
+  c(x_train, x_segment, y_train) %<-% tokenize_fun(data)
+  return(list(x_train, x_segment, y_train))
+}
+```
+
+## Load dataset
+
+As we have already written the preprocess function, at first, it will read data, then add zeros and encode words into indices. Hence, we will have 3 output files:
```

Comment:
```suggestion
The way we have written the preprocess function, at first, it will read data, then add zeros and encode words into indices. Hence, we will have 3 output files:
```